spring-integration-file/src/main/java/org/springframework/integration/file/FileWritingMessageHandler.java

```diff
@@ -481,7 +481,25 @@ public void stop() {
 				this.flushTask = null;
 			}
 		}
-		new Flusher().run();
+		Flusher flusher = new Flusher();
+		flusher.run();
+		boolean needInterrupt = this.fileStates.size() > 0;
+		int n = 0;
+		while (n++ < 10 && this.fileStates.size() > 0) {
+			try {
+				Thread.sleep(1);
+			}
+			catch (InterruptedException e) {
+				// cancel the interrupt
+			}
+			flusher.run();
+		}
+		if (this.fileStates.size() > 0) {
+			this.logger.error("Failed to flush after multiple attempts, while stopping: " + this.fileStates.keySet());
+		}
+		if (needInterrupt) {
+			Thread.currentThread().interrupt();
```

Comment:
Why do you interrupt the thread who calls this `stop()`?

Reply:
Because if we don't complete the flush in the first attempt, the user's thread was interrupted; we can't "eat" that interrupt so after we are complete (or give up) we re-interrupt the thread. That said, I see a problem; pushing another commit...

Reply:
Forget that; all is ok; we reset the interrupt flag each time in the loop.